apollo-configservice/src/test/java/com/ctrip/framework/apollo/configservice/service/config/ConfigServiceWithCacheAndCacheKeyIgnoreCaseTest.java

```diff
@@ -0,0 +1,378 @@
+        .toLowerCase();
+    someNotificationMessages = new ApolloNotificationMessages();
+  }
+
+  @Test
+  public void testFindActiveOne() {
+    long someId = 1;
+
+    when(releaseService.findActiveOne(someId)).thenReturn(someRelease);
+
+    assertEquals(someRelease,
+        configServiceWithCache.findActiveOne(someId, someNotificationMessages));
+    verify(releaseService, times(1)).findActiveOne(someId);
+  }
+
+  @Test
+  public void testFindActiveOneWithSameIdMultipleTimes() {
+    long someId = 1;
+
+    when(releaseService.findActiveOne(someId)).thenReturn(someRelease);
+
+    assertEquals(someRelease,
+        configServiceWithCache.findActiveOne(someId, someNotificationMessages));
+    assertEquals(someRelease,
+        configServiceWithCache.findActiveOne(someId, someNotificationMessages));
+    assertEquals(someRelease,
+        configServiceWithCache.findActiveOne(someId, someNotificationMessages));
+
+    verify(releaseService, times(1)).findActiveOne(someId);
+  }
+
+  @Test
+  public void testFindActiveOneWithMultipleIdMultipleTimes() {
+    long someId = 1;
+    long anotherId = 2;
+    Release anotherRelease = mock(Release.class);
+
+    when(releaseService.findActiveOne(someId)).thenReturn(someRelease);
+    when(releaseService.findActiveOne(anotherId)).thenReturn(anotherRelease);
+
+    assertEquals(someRelease,
+        configServiceWithCache.findActiveOne(someId, someNotificationMessages));
+    assertEquals(someRelease,
+        configServiceWithCache.findActiveOne(someId, someNotificationMessages));
+
+    assertEquals(anotherRelease,
+        configServiceWithCache.findActiveOne(anotherId, someNotificationMessages));
+    assertEquals(anotherRelease,
+        configServiceWithCache.findActiveOne(anotherId, someNotificationMessages));
+
+    verify(releaseService, times(1)).findActiveOne(someId);
+    verify(releaseService, times(1)).findActiveOne(anotherId);
+  }
+
+  @Test
+  public void testFindActiveOneWithReleaseNotFoundMultipleTimes() {
+    long someId = 1;
+
+    when(releaseService.findActiveOne(someId)).thenReturn(null);
+
+    assertNull(configServiceWithCache.findActiveOne(someId, someNotificationMessages));
+    assertNull(configServiceWithCache.findActiveOne(someId, someNotificationMessages));
+    assertNull(configServiceWithCache.findActiveOne(someId, someNotificationMessages));
+
+    verify(releaseService, times(1)).findActiveOne(someId);
+  }
+
+  @Test
+  public void testFindLatestActiveRelease() {
+    when(releaseMessageService.findLatestReleaseMessageForMessages(Lists.newArrayList(lowerCaseSomeKey)))
+        .thenReturn(someReleaseMessage);
+    when(releaseService.findLatestActiveRelease(
+        matches(Pattern.compile(someAppId, Pattern.CASE_INSENSITIVE)),
+        matches(Pattern.compile(someClusterName, Pattern.CASE_INSENSITIVE)),
+        matches(Pattern.compile(someNamespaceName, Pattern.CASE_INSENSITIVE))
```

Comment:
How about replacing it with this method?

```Java
private String matchesCaseInsensitive(final String regex) {
    return matches(Pattern.compile(regex, Pattern.CASE_INSENSITIVE));
}
```

Reply:
good